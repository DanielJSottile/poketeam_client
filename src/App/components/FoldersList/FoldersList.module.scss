@import '../../../index.scss';

.submit {
  display: inline-block;
  text-shadow: none;
  font-weight: bolder;
  background-color: $super-light-cream;
  font-size: 16px;
  color: $med-dark-grey;
  border-radius: 4px;
  padding: 3px;
  margin: 3px;
  text-decoration: none;
}

.buttons {
  margin: 2rem;
  display: flex;
  flex-direction: column;
}

.add-folder-button,
.copy-text-button,
.share-link,
.submit,
.edit,
.delete {
  text-align: center;
  font-family: $body-font;
  padding: 5px;
  border-radius: 40px;
  border: 5px double $med-dark-grey;
}

.copy-text-button {
  font-size: 1rem;
  margin: auto;
}

.share-link {
  width: max-content;
  margin: auto;
}

.submit,
.edit,
.delete {
  width: 9.375rem;
  margin: auto;
}

.add-folder-button {
  width: 200px;
  font-size: 1.5rem;
  margin-left: auto;
  margin-right: auto;
  margin-top: 1rem;
  margin-bottom: 0;
}

.folders {
  display: block;
}

.submit {
  color: green;
}

// .submit:disabled {
//   color: $pokeball-red;
// }

// .submit:hover {
//   background-color: $light-yellow;
//   transition: 0.3s;
// }

.pokeball-div {
  margin: 10px;
}

.hint {
  font-size: 14px;
  letter-spacing: 6px;
}

.share-input {
  width: 12.1875rem;
  padding: 0.3125rem;
}

@media (min-width: 700px) {
  .buttons {
    margin: 2rem 0;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    .inner-buttons {
      display: flex;
      flex-direction: row;
      .label {
        padding-right: 1rem;
      }
    }
  }
}

@media (min-width: 800px) {
  .export-folder,
  .folder-import {
    width: 85%;
    margin: auto;
  }
}
