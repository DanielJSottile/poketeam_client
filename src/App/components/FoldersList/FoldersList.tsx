import React, { Fragment, useContext, useState } from 'react';
import { Link } from 'react-router-dom';
import Folder from '../Folder/Folder';
import Input from '../Input/Input';
import TextArea from '../TextArea/TextArea';
import Button from '../Button/Button';
import PokeballLoader from '../Loaders/PokeballLoader/PokeballLoader';
import LoadingBlack from '../Loaders/LoadingBlack/LoadingBlack';
import GeneralContext from '../../contexts/GeneralContext';
import showdownFolderGenerate from '../../functions/generateFolder';
import styles from './FoldersList.module.scss';

const FoldersList: React.FC = () => {
  const GenCon = useContext(GeneralContext);

  const [state, setState] = useState({
    editClicked: false,
    deleteClicked: false,
    copySuccess: false,
  });

  const handleEditExpand = () => {
    setState((oldVals) => ({ ...oldVals, editClicked: !state.editClicked }));
  };

  const handleDeleteExpand = () => {
    setState((oldVals) => ({
      ...oldVals,
      deleteClicked: !state.deleteClicked,
    }));
  };

  const removeCopySuccess = (): any => {
    setState((oldVals) => ({ ...oldVals, copySuccess: false }));
  };

  const textArea: any = React.useRef(null);

  const copyCodeToClipboard = (): any => {
    textArea.current.select();
    document.execCommand('copy'); // this seems to not work
    const text = textArea.current.defaultValue;
    navigator.clipboard.writeText(text); // this seems to work!
    setState((oldVals) => ({ ...oldVals, copySuccess: true }));
  };

  /* --------------------- */

  /* Set Up Common Definitions to be 
  Used in Different views */

  const {
    userFolders,
    userTeams,
    userSets,
    folderAddClicked,
    currentClickedFolder,
    handleFolderAddClickExpand,
  } = GenCon;

  const folderList = userFolders.map((folder: any, i) => {
    return <Folder key={i} id={folder.id} folder_name={folder.folder_name} />;
  });

  const folderTeams = userTeams.filter(
    (team) => team.folder_id === currentClickedFolder.id
  );

  const input = folderTeams.map((team) => {
    const teamSets = userSets.filter((set) => set.team_id === team.id);
    const teamName: any = team.team_name;
    return { [teamName]: teamSets };
  });

  const renderExpanded = (): JSX.Element => {
    const {
      newFolderName,
      newFolderImport,
      validateNewFolderImport,
      setNewFolderContents,
      setNewFolderName,
      handlePostNewFolder,
      validateNewFolderName,
    } = GenCon;

    return (
      <form>
        <div>
          <Input
            inputHasError
            htmlFor="foldername"
            label="Folder Name:"
            validationCallback={validateNewFolderName()}
            placeholder="e.g. Good Teams"
            type="text"
            name="foldername"
            id="foldername"
            value={newFolderName.value}
            onChangeCallback={(e) => setNewFolderName(e.target.value)}
          />
          <TextArea
            containerClass={styles['folder-import']}
            textAreaHasError
            isError={!!newFolderImport.value}
            validationCallback={validateNewFolderImport()}
            htmlFor="folder-import"
            label="Import Showdown Folder:"
            placeholder="Optionally Import a proper Pokemon Showdown Folder Here And It Will Fill Out The Entire Folder!"
            name="folder-import"
            id="team-import-1"
            value={newFolderImport.value}
            onChangeCallback={(e) => setNewFolderContents(e.target.value)}
          />
        </div>
        <Button
          type="submit"
          buttonClass={styles['submit']}
          disabled={validateNewFolderName() || validateNewFolderImport()}
          onClickCallback={(e) => {
            e.preventDefault();
            handlePostNewFolder();
          }}
        >
          Submit <i className="far fa-check-circle"></i>
        </Button>
      </form>
    );
  };

  const renderEditExpand = (): JSX.Element => {
    const {
      newFolderName,
      setNewFolderName,
      handleEditFolder,
      validateNewFolderName,
      handleCurrentFolderClicked,
      currentClickedFolder,
    } = GenCon;

    return (
      <form>
        <Input
          inputHasError
          htmlFor="foldername"
          label="Edit Folder Name:"
          validationCallback={validateNewFolderName()}
          placeholder="e.g. Good Teams"
          type="text"
          name="foldername"
          id="foldername"
          value={newFolderName.value}
          onChangeCallback={(e) => setNewFolderName(e.target.value)}
        />
        <Button
          type="submit"
          buttonClass={styles['submit']}
          disabled={validateNewFolderName()}
          onClickCallback={(e) => {
            e.preventDefault();
            handleEditFolder();
            handleEditExpand();
            handleCurrentFolderClicked(
              newFolderName.value,
              currentClickedFolder.id
            );
          }}
        >
          Submit <i className="far fa-check-circle"></i>
        </Button>
      </form>
    );
  };

  const renderDeleteExpand = (): JSX.Element => {
    const { handleDeleteFolder } = GenCon;

    return (
      <div>
        <p>Are You Sure You'd Like to Delete this Folder?</p>
        <Button
          onClickCallback={() => {
            handleDeleteFolder();
            handleDeleteExpand();
          }}
        >
          Yes <i className="fas fa-thumbs-up"></i>
        </Button>
        <Button onClickCallback={() => handleDeleteExpand()}>
          No <i className="fas fa-thumbs-down"></i>
        </Button>
      </div>
    );
  };

  return (
    <Fragment>
      <section className={styles['folders-list']}>
        <h3>Folders:</h3>
        <div className={styles['folders']}>
          {folderList.length > 0 ? (
            folderList
          ) : (
            <div className={styles['pokeball-div']}>
              <PokeballLoader />
              <LoadingBlack />
              <h3 className={styles['hint']}>
                (Hint: There May Be No Folders! Click Below to Start!)
              </h3>
            </div>
          )}
        </div>
        <div>
          <Button onClickCallback={() => handleFolderAddClickExpand()}>
            New Folder <i className="fas fa-folder-plus"></i>
          </Button>
          {folderAddClicked ? renderExpanded() : null}
        </div>
        <div>
          <span>{`Current Folder: ${currentClickedFolder.value}`}</span>
          {currentClickedFolder.value ? (
            <div>
              <div className={styles['export-team']}>
                {state.copySuccess ? (
                  <div className={styles['copied']}>Copied to Clipboard!!</div>
                ) : null}
                <div>
                  <Button
                    onClickCallback={() => {
                      copyCodeToClipboard();
                      setTimeout(removeCopySuccess, 3000);
                    }}
                  >
                    Copy Text
                  </Button>
                  <Link
                    to={{
                      pathname: `/share/user/folder/${currentClickedFolder.id}`,
                      state: {
                        folders: userFolders,
                        teams: userTeams,
                        sets: userSets,
                        input: input,
                      },
                    }}
                    target="_blank"
                  >
                    Share This Folder! <i className="fas fa-share-square"></i>
                  </Link>
                  <Input
                    inputHasError={false}
                    disabled
                    type="text"
                    readOnly
                    value={`poketeams.now.sh/share/user/folder/${currentClickedFolder.id}`}
                  />
                </div>
                <label htmlFor="edit-team">
                  Export Folder: <i className="fas fa-download"></i>
                </label>
                <TextArea
                  textAreaHasError={false}
                  ref={textArea}
                  disabled
                  readOnly
                  name="export-folder"
                  id={`export-folder-${currentClickedFolder.id}`}
                  value={showdownFolderGenerate(
                    currentClickedFolder.value,
                    input
                  )}
                />
              </div>
              <Button onClickCallback={() => handleEditExpand()}>
                <i className="fas fa-edit"></i> Edit
              </Button>
              <Button onClickCallback={() => handleDeleteExpand()}>
                Delete <i className="fas fa-trash-alt"></i>
              </Button>
            </div>
          ) : null}
        </div>
        <div>
          {state.editClicked ? renderEditExpand() : null}
          {state.deleteClicked ? renderDeleteExpand() : null}
        </div>
      </section>
    </Fragment>
  );
};

export default FoldersList;
