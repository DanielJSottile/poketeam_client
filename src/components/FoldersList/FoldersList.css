.submit {
  display: inline-block;
  text-shadow: none;
  font-weight: bolder;
  background-color: white;
  font-size: 16px;
  color: #444;
  border-radius: 4px;
  padding: 3px;
  margin: 3px;
  text-decoration: none;
}

.submit{
  color: green;
}

.submit:disabled {
  color: red;
}

.submit:hover {
  background-color: rgb(255, 245, 151);
  transition: 0.3s;
}

.pokeball-div {
  margin: 10px;
}

.hint {
  font-size: 14px;
  letter-spacing: 6px;
}