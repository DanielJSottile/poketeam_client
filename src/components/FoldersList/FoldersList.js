import React, { Component , Fragment} from 'react';
import UserContext from '../../contexts/UserContext';
import Folder from '../Folder/Folder';

export default class FoldersList extends Component {

  static contextType = UserContext;

  renderExpanded() {

    const {
      newFolderName,
      setNewFolderName,
      handlePostNewFolder,
      validateNewFolderName
    } = this.context;

    return (
      <form>
        <div>
          <label htmlFor="foldername">Folder Name:</label>
          {<p className="error">{validateNewFolderName()}</p>}
          <input placeholder="e.g. Good Teams" type="text" name="foldername" id="foldername" value={newFolderName.value} onChange={e => setNewFolderName(e.target.value)}/>
        </div>
        <button type="submit"
        disabled={
          validateNewFolderName()
        }
        onClick={(e) => {
          e.preventDefault();
          handlePostNewFolder();
        }}>Submit</button>
      </form>
    )
  }

  render() {

    const {
      userFolders, 
      folderAddClicked, 
      currentClickedFolder,
      handleFolderAddClickExpand,
    } = this.context;

    const folderList = userFolders.map((folder, i) => {
      return <Folder key={i} id={folder.id} folder_name={folder.folder_name}/>
    });

    return (
      <Fragment>
        <section className="folders-list">
          <h3>Folders:</h3>
          <div>
            {folderList.length > 0 ? folderList : <h3>None! Click Below to Make a New Folder!</h3>}
          </div>
          <div>
            <button onClick={() => handleFolderAddClickExpand()}>New Folder +</button>
            {folderAddClicked ? this.renderExpanded() : null}
          </div>
        <div>
          <span>{`Current Folder: ${currentClickedFolder.value}`}</span>
        </div>
      </section>
    </Fragment>
    );
  };
};
