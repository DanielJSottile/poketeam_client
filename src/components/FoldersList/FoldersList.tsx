import React, { Fragment, useContext, useState } from 'react';
import { Link } from 'react-router-dom';
import GeneralContext from '../../contexts/GeneralContext';
import Folder from '../Folder/Folder';
import PokeballLoader from '../Loaders/PokeballLoader/PokeballLoader';
import LoadingBlack from '../Loaders/LoadingBlack/LoadingBlack';
import showdownFolderGenerate from '../../functions/generateFolder';
import './FoldersList.css';

const FoldersList = (props: any) => {

  const GenCon = useContext(GeneralContext);

  const [state, setState] = useState(
    {
    editClicked: false,
    deleteClicked: false,
    copySuccess: false,
  }
  );

  const handleEditExpand = () => {
    setState(oldVals => ({...oldVals, editClicked: !state.editClicked}));
  }

  const handleDeleteExpand = () => {
    setState(oldVals => ({...oldVals, deleteClicked: !state.deleteClicked}));
  }

  const removeCopySuccess = (): any => {
    setState(oldVals => ({...oldVals, copySuccess: false}))
  }

  // copy to clipboard

  const textArea: any = React.useRef(null);

  const copyCodeToClipboard = (): any => {
    textArea.current.select()
    document.execCommand('copy') // this seems to not work
    const text = textArea.current.defaultValue
    navigator.clipboard.writeText(text) // this seems to work!
    setState(oldVals => ({...oldVals, copySuccess: true}))
  }

  const renderExpanded = (): JSX.Element => {

    const {
      newFolderName,
      newFolderImport,
      validateNewFolderImport,
      setNewFolderContents,
      setNewFolderName,
      handlePostNewFolder,
      validateNewFolderName
    } = GenCon;

    return (
      <form>
        <div>
          <label htmlFor="foldername">Folder Name:</label>
          {<p className="error-validate shake-horizontal">{validateNewFolderName()}</p>}
          <input placeholder="e.g. Good Teams" type="text" name="foldername" id="foldername" value={newFolderName.value} onChange={e => setNewFolderName(e.target.value)}/>
          <div className="folder-import">
          <label htmlFor="folder-import">Import Showdown Folder:</label>
          {newFolderImport.value !== "" && <p className="error-validate shake-horizontal">{validateNewFolderImport()}</p>}
          <textarea placeholder="Optionally Import a proper Pokemon Showdown Folder Here And It Will Fill Out The Entire Folder!" name="folder-import" id="team-import-1" value={newFolderImport.value} onChange={e => setNewFolderContents(e.target.value)}></textarea>
          </div>
        </div>
        <button type="submit"
        className="submit"
        disabled={
          validateNewFolderName() || validateNewFolderImport()
        }
        onClick={(e) => {
          e.preventDefault();
          handlePostNewFolder();
        }}>Submit <i className="far fa-check-circle"></i></button>
      </form>
    )
  }

  const renderEditExpand = (): JSX.Element => {

    const {
      newFolderName,
      setNewFolderName,
      handleEditFolder,
      validateNewFolderName,
      handleCurrentFolderClicked,
      currentClickedFolder
    } = GenCon;

    return (
      <form>
        <div>
          <label htmlFor="foldername">Edit Folder Name:</label>
          {<p className="error-validate shake-horizontal">{validateNewFolderName}</p>}
          <input placeholder="e.g. Good Teams" type="text" name="foldername" id="foldername" value={newFolderName.value} onChange={e => setNewFolderName(e.target.value)}/>
        </div>
        <button type="submit"
        className="submit"
        disabled={
          validateNewFolderName()
        }
        onClick={(e) => {
          e.preventDefault();
          handleEditFolder();
          handleEditExpand();
          handleCurrentFolderClicked(newFolderName.value, currentClickedFolder.id)
        }}>Submit <i className="far fa-check-circle"></i></button>
      </form>
    )
  }

  const renderDeleteExpand = (): JSX.Element => {

    const {
      handleDeleteFolder,
    } = GenCon;

    return (
      <div>
        <p>Are You Sure You'd Like to Delete this Folder?</p> 
        <button onClick={() => {
          handleDeleteFolder();
          handleDeleteExpand();
        }}>Yes <i className="fas fa-thumbs-up"></i></button>
        <button onClick={() => handleDeleteExpand()}>No <i className="fas fa-thumbs-down"></i></button>
      </div> 
    )
  }

  const {
    userFolders,
    userTeams,
    userSets,
    folderAddClicked, 
    currentClickedFolder,
    handleFolderAddClickExpand,
  } = GenCon;

  const folderList = userFolders.map((folder: any, i) => {
    return <Folder key={i} id={folder.id} folder_name={folder.folder_name}/>
  });

  const folderTeams = userTeams.filter((team) => team.folder_id === currentClickedFolder.id)
  
  const input = folderTeams.map((team) => {
    const teamSets = userSets.filter((set) => set.team_id === team.id)
    const teamName: any = team.team_name;
    return {[teamName] : teamSets}
  })

  // here we make the input for the generator function...

  return (
    <Fragment>
      <section className="folders-list">
        <h3>Folders:</h3>
        <div>
          {folderList.length > 0 ? folderList : 
          <div className="pokeball-div">
          <PokeballLoader/>
          <LoadingBlack/>
          <h3 className="hint">(Hint: There May Be No Folders! Click Below to Start!)</h3>
        </div>}
        </div>
        <div>
          <button onClick={() => handleFolderAddClickExpand()}>New Folder <i className="fas fa-folder-plus"></i></button>
          {folderAddClicked ? renderExpanded() : null}
        </div>
      <div>
        <span>{`Current Folder: ${currentClickedFolder.value}`}</span>
        {currentClickedFolder.value ? 
          <div>
            <div className="export-team">
              {
                state.copySuccess ?
                <div className='copied'>
                  Copied to Clipboard!!
                </div> : null
              }
              <div>
                <button onClick={() => {
                  copyCodeToClipboard()
                  setTimeout(removeCopySuccess, 3000)
                }}>Copy Text</button>
                <Link to={{
                  pathname: `/share/folder/${currentClickedFolder.id}`,
                  state: {folders: userFolders}}} target="_blank" >Share This Folder! <i className="fas fa-share-square"></i></Link>
                <input disabled type="text" readOnly value={`poketeams.now.sh/share/folder/${currentClickedFolder.id}`}/>
              </div>
                <label htmlFor="edit-team">Export Folder: <i className="fas fa-download"></i></label>
                <textarea ref={textArea} disabled readOnly name="export-folder" id={`export-folder-${currentClickedFolder.id}`} value={showdownFolderGenerate(currentClickedFolder.value, input)}/>
              </div>
        
            <button onClick={() => handleEditExpand()}><i className="fas fa-edit"></i> Edit</button>
            <button onClick={() => handleDeleteExpand()}>Delete <i className="fas fa-trash-alt"></i></button>
          </div> : null}
      </div>
      <div>
          {state.editClicked ? renderEditExpand() : null}
          {state.deleteClicked ? renderDeleteExpand() : null}
      </div>
    </section>
  </Fragment>
  );
}

export default FoldersList;
