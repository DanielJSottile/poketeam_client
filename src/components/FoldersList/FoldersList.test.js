import React from 'react';
import ReactDOM from 'react-dom';
//import { render } from '@testing-library/react';
import FoldersList from './FoldersList';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<FolderList/>', () => {
  
  it('renders without crashing', () => {
    const wrapper = shallow(<FoldersList/>)
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});