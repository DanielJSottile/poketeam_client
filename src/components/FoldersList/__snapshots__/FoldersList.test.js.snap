// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FolderList/> renders without crashing 1`] = `<FoldersList />`;
